geo/src/algorithm/bool_ops/mod.rs

```diff
@@ -125,6 +132,83 @@ pub enum OpType {
     Xor,
 }
 
+/// Efficient [BooleanOps::union] of adjacent / overlapping geometries
+///
+/// For geometries with a high degree of overlap or adjacency
+/// (for instance, merging a large contiguous area made up of many adjacent polygons)
+/// this method will be orders of magnitude faster than a manual iteration and union approach.
+pub trait UnaryUnion {
+    type Scalar: BoolOpsNum;
+
+    /// Construct a tree of all the input geometries and progressively union them from the "bottom up"
+    ///
+    /// This is considerably more efficient than using e.g. `fold()` over an iterator of Polygons.
+    /// # Examples
+    ///
+    /// ```
+    /// use geo::{BooleanOps, UnaryUnion};
+    /// use geo::{MultiPolygon, polygon};
+    /// let poly1 = polygon![
+    ///     (x: 0.0, y: 0.0),
+    ///     (x: 4.0, y: 0.0),
+    ///     (x: 4.0, y: 4.0),
+    ///     (x: 0.0, y: 4.0),
+    ///     (x: 0.0, y: 0.0),
+    /// ];
+    /// let poly2 = polygon![
+    ///     (x: 4.0, y: 0.0),
+    ///     (x: 8.0, y: 0.0),
+    ///     (x: 8.0, y: 4.0),
+    ///     (x: 4.0, y: 4.0),
+    ///     (x: 4.0, y: 0.0),
+    /// ];
+    /// let merged = &poly1.union(&poly2);
+    /// let mp = MultiPolygon(vec![poly1, poly2]);
+    /// // A larger single rectangle
+    /// let combined = mp.unary_union();
+    /// assert_eq!(&combined, merged);
+    /// ```
+    fn unary_union(&self) -> MultiPolygon<Self::Scalar>;
+}
+
+// This function carries out a full post-order traversal of the tree, building up MultiPolygons from inside to outside.
+// Though the operation is carried out via fold() over the tree iterator, there are two actual nested operations:
+// "fold" operations on leaf nodes build up output MultiPolygons by adding Polygons to them via union and
+// "reduce" operations on parent nodes combine these output MultiPolygons from leaf operations by recursion
+fn bottom_up_fold_reduce<T, S, I, F, R>(
+    tree: &RTree<T>,
+    mut init: I,
+    mut fold: F,
+    mut reduce: R,
+) -> S
+where
+    T: RTreeObject,
+    I: FnMut() -> S,
+    F: FnMut(S, &T) -> S,
+    R: FnMut(S, S) -> S,
+{
+    fn inner<T, S, I, F, R>(parent: &ParentNode<T>, init: &mut I, fold: &mut F, reduce: &mut R) -> S
```

Comment:
In my experience, recursive algorithms like this one can benefit from grouping those parameters which are constant over the whole algorithm to reduce the overhead of the recursive calls. So something like

```rust
struct Ops<I, F, R> {
  init: I,
  fold: F,
  reduce: R,
}

fn inner<T, S, I, F, R>(ops: &mut Ops<I, F, R>, parent: &ParentNode<T>) -> S /* .. */

inner(&mut Ops { init, fold, reduce }, tree.root())
```

Reply:
Done in ba95b2c86fa43d88dad32c48167c12975d64cd36